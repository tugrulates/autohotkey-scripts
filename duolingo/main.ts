import { Command } from "@cliffy/command";
import { Table } from "@cliffy/table";
import { pool } from "@tugrulates/internal/async";
import { Config } from "@tugrulates/internal/cli";
import { DuolingoClient } from "./client.ts";
import { LEAGUES } from "./data.ts";
import {
  engageWithCard,
  followLeagueUsers,
  getEmoji,
  getLeagueUserEmoji,
} from "./interaction.ts";
import type { FeedCard, League } from "./types.ts";

let username: Config | undefined;
let token: Config | undefined;

/** Duolingo client built from common CLI options. */
export async function getClient(): Promise<DuolingoClient> {
  if (!username || !token) {
    throw new Error("Username and token not configured.");
  }
  return new DuolingoClient(await username.get(), await token.get());
}

/**
 * Fetches and organizes follow information.
 *
 * @returns Users who are followed, users who follow, and their difference sets.
 */
async function getFollows() {
  const client = await getClient();
  const [following, followers] = await Promise.all([
    client.getFollowing(),
    client.getFollowers(),
  ]);
  return {
    following,
    followers,
    dontFollowBack: following.filter(({ userId }) =>
      !followers.some((user) => user.userId === userId)
    ),
    notFollowingBack: followers.filter(({ userId }) =>
      !following.some((user) => user.userId === userId)
    ),
  };
}

/**
 * Outputs the league to the console.
 *
 * @param league League to output.
 */
async function outputLeague(league: League) {
  const client = await getClient();
  const following = await client.getFollowing();
  const tier = LEAGUES[league.tier];
  new Table()
    .header([tier.emoji, tier.name])
    .body(
      league.rankings.map((user, index) => [
        `${index + 1}.`,
        `${user.display_name} ${getLeagueUserEmoji(user)}`,
        following ? "👤" : "",
        `${user.score.toString()} XP`,
      ]),
    )
    .columns([{ align: "right" }, {}, {}, { align: "right" }])
    .render();
}

/**
 * Returns the display summary of the card.
 *
 * @param card Card to get the summary for.
 * @returns Display summary of the card.
 */
function getSummary(card: FeedCard): string {
  return card.header?.replace(/<[^>]+>/g, "") ??
    `${card.displayName} ${card.body.toLowerCase()}`;
}

function getFeedCommand() {
  return new Command()
    .description("Prints and interacts with the feed.")
    .example("duolingo feed", "Prints the feed.")
    .example("duolingo feed --engage", "Engages with the feed.")
    .example("duolingo feed --json | jq", "Query JSON over the feed.")
    .option("--engage", "Engage with the feed events.")
    .option("--json", "Output the feed as JSON.")
    .action(async ({ engage, json }) => {
      const client = await getClient();
      const followers = await client.getFollowers();
      const cards = await client.getFeedCards();
      if (json) console.log(JSON.stringify(cards, undefined, 2));
      await pool(
        cards,
        async (card) => {
          if (!engage || await engageWithCard(client, followers, card)) {
            if (!json) console.log(`${getEmoji(card)} ${getSummary(card)}`);
          }
        },
      );
    });
}

function getFollowsCommand() {
  return new Command()
    .description("Prints and manages follower information on Duolingo.")
    .example("duolingo follows", "Prints follow counts.")
    .example("duolingo follows --follows", "Follow users who follow.")
    .example("duolingo follows --unfollow", "Unfollow users who dont' follow.")
    .example("duolingo follows --follow --unfollow", "Matches both lists.")
    .example("duolingo follows --json", "Outputs JSON of follower information.")
    .example(
      "duolingo follows --json | jq",
      "Query JSON for follower information.",
    )
    .example(
      "duolingo follows --json | jq '.dontFollowBack[].username'",
      "List users who are followed but don't follow back.",
    )
    .example(
      "duolingo follows --json | jq '.notFollowingBack[].username'",
      "List users who follow but are not followed back.",
    )
    .option("--follow", "Follow users who follow.")
    .option("--unfollow", "Unfollow users who don't follow.")
    .option("--json", "Output the follower information as JSON.")
    .action(async ({ follow, unfollow, json }) => {
      const client = await getClient();
      let result = await getFollows();

      if (follow || unfollow) {
        if (follow) {
          await pool(
            result.notFollowingBack,
            async (user) => {
              await client.followUser(user.userId);
              if (!json) console.log(`✅ Followed ${user.username}.`);
            },
          );
        }
        if (unfollow) {
          await pool(
            result.dontFollowBack,
            async (user) => {
              await client.unfollowUser(user.userId);
              if (!json) console.log(`❌ Unfollowed ${user.username}.`);
            },
          );
        }
        result = await getFollows();
      }

      if (json) console.log(JSON.stringify(result, undefined, 2));
      else {
        console.log(`👤 Following ${result.following.length} people.`);
        console.log(`👤 Followed by ${result.followers.length} people.`);
      }
    });
}

function getLeagueCommand() {
  return new Command()
    .description("Prints and interacts with the current Duolingo league.")
    .example("duolingo league", "Prints the league.")
    .example("duolingo league --follow", "Follows users in the league.")
    .example("duolingo league --json | jq", "Query JSON over the league.")
    .option("--follow", "Follow users in the league.")
    .option("--json", "Output the league as JSON.")
    .action(async ({ follow, json }) => {
      const client = await getClient();
      const league = await client.getLeague();
      if (follow) await followLeagueUsers(client, league.rankings);
      if (json) console.log(JSON.stringify(league, undefined, 2));
      else await outputLeague(league);
    });
}

async function getCommand() {
  const command = new Command()
    .name("duolingo")
    .description("Interact with Duolingo.")
    .example("duolingo --username <username> --token <token>", "Configure.")
    .example("duolingo --clear", "Clear the cached configuration.")
    .usage("<command> [options]")
    .globalOption(
      "--username <username:string>",
      "Username.",
      await username?.option(),
    )
    .globalOption(
      "--token <token:string>",
      "JWT token.",
      await token?.option(),
    )
    .option("--clear", "Clear the cached configuration.", {
      standalone: true,
      action: async () => {
        await username?.clear();
        await token?.clear();
      },
    })
    .action((): void => command.showHelp())
    .command("feed", getFeedCommand())
    .command("follows", getFollowsCommand())
    .command("league", getLeagueCommand());
  return command;
}

/** CLI entrypoint. */
export async function main() {
  username = new Config("username");
  token = new Config("token", { secret: true });
  const command = await getCommand();
  await command.parse();
}
