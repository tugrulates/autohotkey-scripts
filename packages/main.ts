import { Command } from "@cliffy/command";
import { Table } from "@cliffy/table";
import {
  bumpVersion,
  findCommitsSinceLastRelease,
  getPackage,
  getWorkspacePackages,
  setPackage,
} from "@tugrulates/internal/package";

function getListCommand() {
  return new Command()
    .description("Lists all packages in the workspace.")
    .example("packages list", "List all Deno packages.")
    .action(async () => {
      const packages = await getWorkspacePackages(".");
      new Table()
        .body(
          packages.map((
            pkg,
          ) => ["ðŸ“¦", pkg.directory, pkg.config.name, pkg.config.version]),
        )
        .render();
    });
}

function getBumpCommand() {
  return new Command()
    .description("Prints a list of active and high quality users on 500px.")
    .example("packages bump", "Updates versions for packages with a change.")
    .example("packages bump --dry-run", "Calculate updates without changes.")
    .option("--dry-run", "Calculate updates without changes.")
    .action(async ({ dryRun }) => {
      const packages = await getWorkspacePackages(".");
      const updates = (await Promise.all(
        packages.filter((pkg) => pkg.config.name && pkg.config.version).map(
          async (pkg) => {
            const commits = await findCommitsSinceLastRelease(pkg);
            const version = bumpVersion(pkg, commits);
            return {
              ...pkg,
              newVersion: version,
              commits,
            };
          },
        ),
      )).filter((pkg) => pkg.config.version !== pkg.newVersion);
      if (!dryRun) {
        await Promise.all(
          updates.map((pkg) =>
            setPackage({
              ...pkg,
              config: { ...pkg.config, version: pkg.newVersion },
            })
          ),
        );
      }
      new Table().body(
        updates.map((
          pkg,
        ) => ["ðŸ“Œ", pkg.config.name, pkg.config.version, "ðŸ‘‰", pkg.newVersion]),
      ).render();
    });
}

async function getCommand() {
  const command = new Command()
    .name("packages")
    .description("Manage repository packages.")
    .usage("<command> [options]")
    .version((await getPackage()).config.version ?? "")
    .action((): void => command.showHelp())
    .command("list", getListCommand())
    .command("bump", getBumpCommand());
  return command;
}

/** CLI entrypoint. */
async function main(args: string[]) {
  const command = await getCommand();
  await command.parse(args);
}

if (import.meta.main) await main(Deno.args);
