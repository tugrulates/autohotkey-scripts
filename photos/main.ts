import { colors } from "@cliffy/ansi/colors";
import { Command } from "@cliffy/command";
import { omit } from "@std/collections/omit";
import { expandGlob } from "@std/fs";
import { join } from "@std/path";
import { OPTIONAL_FIELDS, VARIANT_FIELDS } from "./fields.ts";
import { copyExifToVariants, getPhoto } from "./photo.ts";
import type { Photo } from "./types.ts";

const SOURCE_FILE = "source.jpg";

async function isFile(path: string): Promise<boolean> {
  try {
    const info = await Deno.lstat(path);
    return info.isFile;
  } catch {
    return false;
  }
}

async function* getPhotos(photos: string[]): AsyncGenerator<Photo> {
  if (photos.length === 0) {
    for await (const entry of expandGlob(join("*", SOURCE_FILE))) {
      yield await getPhoto(entry.path);
    }
  }
  for (const photo of photos) {
    if (await isFile(photo)) {
      yield (await getPhoto(photo));
    } else if (await isFile(join(photo, SOURCE_FILE))) {
      yield getPhoto(join(photo, SOURCE_FILE));
    }
  }
}

function check(data: Photo) {
  const result = [];
  for (const [field, value] of Object.entries(omit(data, OPTIONAL_FIELDS))) {
    if (!value) result.push(`missing:${field}`);
  }
  for (const size of data.variants) {
    for (const field in omit(size, VARIANT_FIELDS)) {
      if (field in data) result.push(`${size.src}:${field}`);
    }
  }
  if (result.length) return `[${colors.yellow(result.join(", "))}]`;
  return "";
}

function getCommand() {
  return new Command()
    .name("photos")
    .example("photos", "Lists all photos under current directory.")
    .example("photos [photo] --json", "Data for a photo with all variants.")
    .example("photos [photo] --copy", "Copy EXIF data to all variants.")
    .description("Manage photos.")
    .arguments("[photos...:file]")
    .option("--copy", "Copy the EXIF from source JPG to other variants.")
    .option("--json", "Output the EXIF information as JSON.")
    .action(async ({ copy, json }, ...photos) => {
      for await (let photo of getPhotos(photos)) {
        if (copy) {
          await copyExifToVariants(photo);
          photo = await getPhoto(photo.src);
        }
        if (json) console.log(JSON.stringify(photo));
        else console.log(`ðŸ–¼  ${photo.title} ${check(photo)}`);
      }
    });
}

/** CLI entrypoint. */
export async function main(): Promise<void> {
  const command = getCommand();
  await command.parse();
}
