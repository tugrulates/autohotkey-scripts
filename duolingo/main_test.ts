import { assertSnapshot } from "@std/testing/snapshot";
import { fakeConsole, getMockMode, mockFetch } from "@tugrulates/testing";
import { main } from "./main.ts";

// Use ENV variables for recording, but fake credentials for replay.
const CONFIG = [] as string[];
if (getMockMode() === "replay") {
  CONFIG.push("--username", "TugrulAtes", "--token", "TOKEN");
}

Deno.test(
  "duolingo feed",
  async (t) => {
    using console = fakeConsole();
    using _fetch = mockFetch(t);
    await main([...CONFIG, "feed"], { path: ":memory:" });
    await assertSnapshot(t, console.calls);
  },
);

Deno.test.ignore(
  "duolingo feed --engage",
  async (t) => {
    using console = fakeConsole();
    using _fetch = mockFetch(t);
    await main([...CONFIG, "feed", "--engage"], { path: ":memory:" });
    await assertSnapshot(t, console.calls);
  },
);

Deno.test(
  "duolingo follows",
  async (t) => {
    using console = fakeConsole();
    using _fetch = mockFetch(t);
    await main([...CONFIG, "follows"], { path: ":memory:" });
    await assertSnapshot(t, console.calls);
  },
);

Deno.test.ignore(
  "duolingo follows --follow",
  async (t) => {
    using console = fakeConsole();
    using _fetch = mockFetch(t);
    await main([...CONFIG, "follows", "--follow"], { path: ":memory:" });
    await assertSnapshot(t, console.calls);
  },
);

Deno.test.ignore(
  "duolingo follows --unfollow",
  async (t) => {
    using console = fakeConsole();
    using _fetch = mockFetch(t);
    await main([...CONFIG, "follows", "--unfollow"], { path: ":memory:" });
    await assertSnapshot(t, console.calls);
  },
);

Deno.test(
  "duolingo league",
  async (t) => {
    using console = fakeConsole();
    using _fetch = mockFetch(t);
    await main([...CONFIG, "league"], { path: ":memory:" });
    await assertSnapshot(t, console.calls);
  },
);

Deno.test.ignore(
  "duolingo league --follow",
  async (t) => {
    using console = fakeConsole();
    using _fetch = mockFetch(t);
    await main([...CONFIG, "league", "--follow"], { path: ":memory:" });
    await assertSnapshot(t, console.calls);
  },
);
