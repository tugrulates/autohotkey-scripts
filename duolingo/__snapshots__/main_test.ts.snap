export const snapshot = {};

snapshot[`duolingo feed 1`] = `
[
  {
    data: [
      "🎉 Sheri Poisson ONeill completed 5 friends quests in a row!",
    ],
    level: "log",
  },
  {
    data: [
      "💪 Marzenna completed the january challenge!",
    ],
    level: "log",
  },
  {
    data: [
      "💪 Ahmet B completed the january challenge!",
    ],
    level: "log",
  },
  {
    data: [
      "💯 Dani Badaró completed a 500 day streak!",
    ],
    level: "log",
  },
  {
    data: [
      "💪 Budzo11 completed the january challenge!",
    ],
    level: "log",
  },
  {
    data: [
      "🙏 Deniz Can Çalışkan came back to learn dutch after 1 month!",
    ],
    level: "log",
  },
  {
    data: [
      "👍 RightPlaceWrongPersonByRM shared a sentence",
    ],
    level: "log",
  },
  {
    data: [
      "💯 Yoopii earned a total of 20000 xp!",
    ],
    level: "log",
  },
  {
    data: [
      "💪 Norma completed the january challenge!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 HalloI’m Japanese I like game completed 3 friends quests in a row!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 Sharon Bonilla completed a 30 day streak!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 Furkan Mert Şahin completed 5 friends quests in a row!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 ayse completed 30 friends quests in a row!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 Машшка completed 105 friends quests in a row!",
    ],
    level: "log",
  },
  {
    data: [
      "👍 Raizy shared a sentence",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 Naysú completed 3 friends quests in a row!",
    ],
    level: "log",
  },
  {
    data: [
      "👍 Raizy shared a sentence",
    ],
    level: "log",
  },
  {
    data: [
      "👍 Raizy shared a sentence",
    ],
    level: "log",
  },
  {
    data: [
      "💪 Nick completed the january challenge!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 Sheri Poisson ONeill did more than 10 lessons in a day!",
    ],
    level: "log",
  },
  {
    data: [
      "👤 Deborah followed you back!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 Arina completed 3 friends quests in a row!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 Sony completed 20 friends quests in a row!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 Josa Fav completed 3 friends quests in a row!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 Francesco did more than 10 lessons in a day!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 Le meilleur completed 3 friends quests in a row!",
    ],
    level: "log",
  },
  {
    data: [
      "💯 Marzenna learned a total of 2000 words!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 Susi completed 115 friends quests in a row!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 Anna completed 5 friends quests in a row!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 helma auler completed 115 friends quests in a row!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 yonca did more than 10 lessons in a day!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 yonca completed 3 friends quests in a row!",
    ],
    level: "log",
  },
  {
    data: [
      "💪 Ciflorette completed the january challenge!",
    ],
    level: "log",
  },
  {
    data: [
      "💪 Batuhan Bensoy completed the january challenge!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 Machiko completed 10 friends quests in a row!",
    ],
    level: "log",
  },
  {
    data: [
      "👍 RightPlaceWrongPersonByRM shared a sentence",
    ],
    level: "log",
  },
  {
    data: [
      "👍 RightPlaceWrongPersonByRM shared a sentence",
    ],
    level: "log",
  },
  {
    data: [
      "👍 Denver shared a sentence",
    ],
    level: "log",
  },
  {
    data: [
      "💪 João Alfredo completed the january challenge!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 Paula Corina did more than 10 lessons in a day!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 Marlova completed 35 friends quests in a row!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 Raku completed 115 friends quests in a row!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 Duo hi did more than 10 lessons in a day!",
    ],
    level: "log",
  },
  {
    data: [
      "💯 İremnur earned a total of 30000 xp!",
    ],
    level: "log",
  },
  {
    data: [
      "👍 Raizy shared a sentence",
    ],
    level: "log",
  },
  {
    data: [
      "👍 Raizy shared a sentence",
    ],
    level: "log",
  },
  {
    data: [
      "💪 Koro completed the january challenge!",
    ],
    level: "log",
  },
  {
    data: [
      "💯 Magno Henrique completed a 1300 day streak!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 Vyrna23 completed 10 friends quests in a row!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 Duo hi learned a total of 175 words!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 Cris did more than 10 lessons in a day!",
    ],
    level: "log",
  },
  {
    data: [
      "💪 Cris completed the january challenge!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 Gyuri Enikő completed a 30 day streak!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 Olga completed 3 friends quests in a row!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 drumwhacker completed 15 friends quests in a row!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 Denver completed 60 friends quests in a row!",
    ],
    level: "log",
  },
]
`;

snapshot[`duolingo feed --engage 1`] = `
[
  {
    data: [
      "🎉 Carmen did more than 10 lessons in a day!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 KAON0111 completed a 50 day streak!",
    ],
    level: "log",
  },
  {
    data: [
      "💪 Marzenna completed the january challenge!",
    ],
    level: "log",
  },
  {
    data: [
      "💯 Dani Badaró completed a 500 day streak!",
    ],
    level: "log",
  },
  {
    data: [
      "🙏 Deniz Can Çalışkan came back to learn dutch after 1 month!",
    ],
    level: "log",
  },
  {
    data: [
      "👍 RightPlaceWrongPersonByRM shared a sentence",
    ],
    level: "log",
  },
  {
    data: [
      "💯 Yoopii earned a total of 20000 xp!",
    ],
    level: "log",
  },
  {
    data: [
      "💪 Norma completed the january challenge!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 Sharon Bonilla completed a 30 day streak!",
    ],
    level: "log",
  },
  {
    data: [
      "👍 Raizy shared a sentence",
    ],
    level: "log",
  },
]
`;

snapshot[`duolingo follows 1`] = `
[
  {
    data: [
      "👤 Following 353 people.",
    ],
    level: "log",
  },
  {
    data: [
      "👤 Followed by 360 people.",
    ],
    level: "log",
  },
]
`;

snapshot[`duolingo follows --follow 1`] = `
[
  {
    data: [
      "✅ Followed Carmen227854.",
    ],
    level: "log",
  },
  {
    data: [
      "✅ Followed Maga292126..",
    ],
    level: "log",
  },
  {
    data: [
      "✅ Followed MamukaBeri2.",
    ],
    level: "log",
  },
  {
    data: [
      "✅ Followed ywxZ14.",
    ],
    level: "log",
  },
  {
    data: [
      "✅ Followed GinaJara.",
    ],
    level: "log",
  },
  {
    data: [
      "✅ Followed riwo868.",
    ],
    level: "log",
  },
  {
    data: [
      "✅ Followed .Paluten.",
    ],
    level: "log",
  },
  {
    data: [
      "👤 Following 355 people.",
    ],
    level: "log",
  },
  {
    data: [
      "👤 Followed by 360 people.",
    ],
    level: "log",
  },
]
`;

snapshot[`duolingo follows --unfollow 1`] = `
[
  {
    data: [
      "❌ Unfollowed LuigiPicci1.",
    ],
    level: "log",
  },
  {
    data: [
      "❌ Unfollowed maria.68.andrrx.",
    ],
    level: "log",
  },
  {
    data: [
      "👤 Following 353 people.",
    ],
    level: "log",
  },
  {
    data: [
      "👤 Followed by 360 people.",
    ],
    level: "log",
  },
]
`;

snapshot[`duolingo league 1`] = `
[
  {
    data: [
      "🩷  Pearl League              
 1. paulettka Kozdra   1415 XP
 2. Ina                1284 XP
 3. galina              764 XP
 4. winters 😡          758 XP
 5. Michele 😎          645 XP
 6. Karolina            580 XP
 7. Janet Koyumdjian    487 XP
 8. Mika 🇸🇪             464 XP
 9. cloclo              454 XP
10. Alexey SKV          446 XP
11. Malu                445 XP
12. AnFe                436 XP
13. SimonG              396 XP
14. Tuğrul 👀           385 XP
15. Ania                376 XP
16. Blai                355 XP
17. Juraj               351 XP
18. Mona. 🇩🇪            348 XP
19. Emese Szabó         295 XP
20. Yvonne 😺           279 XP
21. Bravo 😎            266 XP
22. Annajc 😡           261 XP
23. Ziv Dor Guetta 🇪🇸   248 XP
24. Bíborka Salamon     232 XP
25. fulgencio           220 XP
26. Angela              217 XP
27. Ujco                179 XP
28. Anne Sophie         176 XP
29. Nathan Lesman       168 XP
30. Wera Bos 🍿         162 XP",
    ],
    level: "log",
  },
]
`;
